From a628f9bee58f7c29e5bc02e592ac6cc0eee22943 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:01:33 -0400
Subject: [usb] ohci: use amd_chipset_type to filter for SB800 prefetch

Message-id: <1410887463-20674-22-git-send-email-dzickus@redhat.com>
Patchwork-id: 95116
O-Subject: [RHEL7 PATCH 021/591] usb: ohci: use amd_chipset_type to filter for SB800 prefetch
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 02c123ee99c793f65af2dbda17d5fe87d448f808
Author: Huang Rui <ray.huang@amd.com>
Date:   Thu Oct 3 23:37:13 2013 +0800

    usb: ohci: use amd_chipset_type to filter for SB800 prefetch

    Commit "usb: pci-quirks: refactor AMD quirk to abstract AMD chipset types"
    introduced a new AMD chipset type to filter AMD platforms with different
    chipsets.

    According to a recent thread [1], this patch updates SB800 prefetch routine
    in AMD PLL quirk. And make it use the new chipset type to represent SB800
    generation.

    [1] http://marc.info/?l=linux-usb&m=138012321616452&w=2

    Signed-off-by: Huang Rui <ray.huang@amd.com>
    Acked-by: Alan Stern <stern@rowland.harvard.edu>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/ohci-pci.c b/drivers/usb/host/ohci-pci.c
index eedf97c..90879e9 100644
--- a/drivers/usb/host/ohci-pci.c
+++ b/drivers/usb/host/ohci-pci.c
@@ -150,28 +150,16 @@ static int ohci_quirk_nec(struct usb_hcd *hcd)
 static int ohci_quirk_amd700(struct usb_hcd *hcd)
 {
  struct ohci_hcd *ohci = hcd_to_ohci(hcd);
- struct pci_dev *amd_smbus_dev;
- u8 rev;
 
  if (usb_amd_find_chipset_info())
   ohci->flags |= OHCI_QUIRK_AMD_PLL;
 
- amd_smbus_dev = pci_get_device(PCI_VENDOR_ID_ATI,
-   PCI_DEVICE_ID_ATI_SBX00_SMBUS, NULL);
- if (!amd_smbus_dev)
-  return 0;
-
- rev = amd_smbus_dev->revision;
-
  /* SB800 needs pre-fetch fix */
- if ((rev >= 0x40) && (rev <= 0x4f)) {
+ if (usb_amd_prefetch_quirk()) {
   ohci->flags |= OHCI_QUIRK_AMD_PREFETCH;
   ohci_dbg(ohci, "enabled AMD prefetch quirk\n");
  }
 
- pci_dev_put(amd_smbus_dev);
- amd_smbus_dev = NULL;
-
  return 0;
 }
 
diff --git a/drivers/usb/host/pci-quirks.c b/drivers/usb/host/pci-quirks.c
index e670882..dfbdd3a 100644
--- a/drivers/usb/host/pci-quirks.c
+++ b/drivers/usb/host/pci-quirks.c
@@ -275,6 +275,14 @@ bool usb_amd_hang_symptom_quirk(void)
 }
 EXPORT_SYMBOL_GPL(usb_amd_hang_symptom_quirk);
 
+bool usb_amd_prefetch_quirk(void)
+{
+ usb_amd_find_chipset_info();
+ /* SB800 needs pre-fetch fix */
+ return amd_chipset.sb_type.gen == AMD_CHIPSET_SB800;
+}
+EXPORT_SYMBOL_GPL(usb_amd_prefetch_quirk);
+
 /*
  * The hardware normally enables the A-link power management feature, which
  * lets the system lower the power consumption in idle states.
diff --git a/drivers/usb/host/pci-quirks.h b/drivers/usb/host/pci-quirks.h
index 820f532..638e88f 100644
--- a/drivers/usb/host/pci-quirks.h
+++ b/drivers/usb/host/pci-quirks.h
@@ -6,6 +6,7 @@ void uhci_reset_hc(struct pci_dev *pdev, unsigned long base);
 int uhci_check_and_reset_hc(struct pci_dev *pdev, unsigned long base);
 int usb_amd_find_chipset_info(void);
 bool usb_amd_hang_symptom_quirk(void);
+bool usb_amd_prefetch_quirk(void);
 void usb_amd_dev_put(void);
 void usb_amd_quirk_pll_disable(void);
 void usb_amd_quirk_pll_enable(void);
-- 
1.7.1